### DISCUSSION TITLE
Hiding materials

### DISCUSSION THREAD

**Discussion by dmkrtz** [google/model-viewer#2893] [Q&A]
### Description

I am using a custom function to hide materials which sets the color of said material to "0,0,0,-1". This worked fine until the last update, forcing me to use 1.8.0 instead of the recently released 1.9.0.
Was this on purpose or am I just abusing a bug? :D 
I mean, it works. See video below. Where it works I used model-viewer@1.8.0, then after pressing reload I changed the src to newest version 1.9.0 and instead of hiding the material it's painted black.

#### Live Demo

https://puu.sh/Ijj97/fb5f20f0f1.mp4

### Version

- model-viewer: v1.9.0

**Comment by elalish** [comment]
Hmm, `[0, 0, 0, -1]` does sound like abusing a bug 😂. It should definitely be possible to hide materials though. Have you tried `[1, 1, 1, 0]`? You'll also need to make sure to set  `alphaMode` to  'MASK' or 'BLEND'.

**Reply by dmkrtz** [reply]
Haha yeah that is the issue. I have some meshes that don't need any `alphaMode`, and by using `0,0,0,-1` I was able to hide any material that I wanted. Now when using `1,1,1,0` I can hide materials that have said `alphaMode` enabled, but it doesn't affect the materials that don't have this enabled.
But this means I have to set an `alphaMode` to each material, because I allow it to hide each material.
Couldn't an actual hideMaterial function be implemented? Like `modelViewer.material[i].hide`?
Because my `-1` trick works perfectly fine and results in exactly what I want.

And setting everything to BLEND results in weird behaviour, see attached image.

![screenshot195928_SD](https://user-images.githubusercontent.com/89161020/137965309-bd757001-e2f9-49db-8e76-fefe734a5902.png)

**Reply by elalish** [reply]
Ah, that's helpful! Can you provide a repro link so I can debug better?

**Reply by dmkrtz** [reply]
sadly I don't use github or anything for my code, as I have to secure my code as good as I can because those models shouldn't be accessible to anyone but my own app.

But the code snippet for a specific button is basically this (hides multiple materials at once):
```
var isChecked = $(this).prop("checked");
var materials = modelViewer.model.materials;
const hideMaterials = [ "rwheel", "sprocket", "rdisc", "fwheel", "fdisc", "chain", "rims"];

if(isChecked) {
for (var i=0,j=materials.length; i<j; i++) {
	if ( hideMaterials.includes(materials[i].name) ) {
		// working in 1.8.0
		// colorString = "1,1,1,-1";
		colorString = "1,1,1,0";

		color = colorString.split(',')
			.map(numberString => parseFloat(numberString));
		
		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
	}
}
} else {
for (var i=0,j=materials.length; i<j; i++) {
	if ( hideMaterials.includes(materials[i].name) ) {
		colorString = "1,1,1,1";

		color = colorString.split(',')
			.map(numberString => parseFloat(numberString));

		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
	}
}
```
This is linked to an element listener and depending on the state of a checkbox it either hides or shows materials from the hideMaterials array. I hope it's somewhat explainable, I'm not a big commenter in my code and I know this could be made easier, but I'm currently working out things and in the end the code isn't very much complicated i think

**Reply by elalish** [reply]
Actually, it's the model I need, since that's where the transparency problem is occurring. Can you share any model that demonstrates the problem? I'd like for the `alphaMode` solution to work well, rather than  bring back a hack.

**Reply by dmkrtz** [reply]
`model` was the keyword I guess.
I just tried to use alpha clip in Blender instead, rather than alpha blend.

This is alpha blend:
![screenshot201954_SD](https://user-images.githubusercontent.com/89161020/137968260-1ad977a1-ca19-46bb-8899-c6699e49e04f.png)

and this is alpha clip:
![screenshot201931_SD](https://user-images.githubusercontent.com/89161020/137968290-83f19846-1760-4f25-8b28-357dc8452aa8.png)

And on both hiding materials with `1,1,1,0` works, just that alpha clip looks how it's supposed to look, lol.

**Reply by elalish** [reply]
Awesome, we'll consider this solved then! Thanks for the descriptions; I'm sure others will benefit too.

**Reply by dmkrtz** [reply]
Thanks for your time and quick replies. Really loving the model-viewer! A lot of fun messing around with it already.

**Reply by Biglow230** [reply]
@dmkrtz 
This is very much what I'm trying to achieve if I'm reading this correctly.

Effectively I have a lot of materials that are in the same coordinates. I'd like to hide all of theme by default except for a base few.

There will be 4 drop down boxes where if you select it will only show the matched value which would be the texture name. I'm also having trouble hiding the material, it's an image texture so Idk if that makes a difference on where I should be targeting.

It has the following properties
alphaCutoff: 0.5
alphaMode: "MASK"
doubleSided: true

Any advice on how to remix this what what I described above?

Major thanks in advance!!

**Reply by elalish** [reply]
Does this example help at all? https://modelviewer.dev/examples/scenegraph/#createTexturesExample

I can't quite tell if you're talking about hiding objects or switching variants.

**Reply by Biglow230** [reply]
It would be to hide and show different materials/objects.

Sorry for the confusion let me provide some context.
I have a few models that are split up as follows. Hat 1, 2, 3 body 1, 2, 3 hand 1, 2, 3 foot 1, 2, 3.
Each item is a different object.

If I export everything into 1 glb, I can see everything in the console. Is there a way to hide the object directly or rather than having 1 massive .glb is there a way to have multiple .glb loaded? I'm assuming the way it's handled in the dirt bike model is by having 1 .glb and targeting the object?

**Reply by elalish** [reply]
So the  trick is that  materials and objects (meshes) are two very different things, and depending which one you want to change, you  need to do it in different ways. Usually the change is to a material, in which case you can just change its textures and other properties as appropriate (and bonus, you only  need to include the first set in  your GLB). If you need to hide whole objects, then they will need their own material  assigned and you can  use the above masking trick to mask/unmask them. The most important part is to set the fourth value of baseColor to zero to make them disappear.

**Reply by Biglow230** [reply]
Ah in this case it's definitely objects with separate materials. I was trying to utilize the example but I wasn't able to see the changes being made in the console.

So if I'm understanding this correctly, the name of the object is in the const and that's being matched with the if else statement on load. If match hide when checked, show if un checked is this correct? 

As was also mentioned a simpler method and or the ability to load multiple models like in this example [https://threejs.org/examples/?q=mul#webgl_animation_multiple](https://threejs.org/examples/?q=mul#webgl_animation_multiple) is potentially in the works?



Would this be similar to the [https://modelviewer.dev/examples/scenegraph/#pickMaterialExample]() ? Only adding a 4th digit to base color?

**Reply by Biglow230** [reply]
@elalish  I'm assuming the structure of the html and Js is as follows.

```
<input type="checkbox" id="checked">




<script type="module">
Const ModelViewerMaterial = document.querySelector("modelviewer")


ModelViewerMaterial.addEventListener("load", () => {

var isChecked = $(this).prop("checked");
var materials = ModelViewerMaterial.model.materials;
const hideMaterials = [ "Material.001", ];

if(isChecked) {
for (var i=0,j=materials.length; i<j; i++) {
	if ( hideMaterials.includes(materials[i].name) ) {
			// working in 1.8.0
					// colorString = "1,1,1,-1";
							colorString = "1,1,1,0";
		color = colorString.split(',')
					.map(numberString => parseFloat(numberString));
		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
			}
			}
			} else {
			for (var i=0,j=materials.length; i<j; i++) {
				if ( hideMaterials.includes(materials[i].name) ) {
						colorString = "1,1,1,1";
		color = colorString.split(',')
					.map(numberString => parseFloat(numberString));
		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
			}
			}

</Script>
```

**Reply by elalish** [reply]
That looks about right; is it  working? You may also need to  change the `materials[i].alphaMode = 'MASK'` as described above.

**Reply by Biglow230** [reply]
@elalish 
Sadly not working. I've included the updated code and the model for reference. 



![image](https://user-images.githubusercontent.com/94942972/144926675-19606cb3-72ac-4f61-8508-f7a912d03604.png)

`modelViewer.addEventListener('load',()=>{

 var isChecked = $(this).prop("checked");
var materials = modelViewer.model.materials;
const hideMaterials = [ "Material0GMObj0.820" ];


if(isChecked) {
for (var i=0,j=materials.length; i<j; i++) {
	if ( hideMaterials.includes(materials[i].name) ) {
			// working in 1.8.0
					// colorString = "1,1,1,-1";
							colorString = "1,1,1,0";
		color = colorString.split(',')
					.map(numberString => parseFloat(numberString));
          modelViewer.model.materials[i].alphaMode = 'MASK'
		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
			}
			}
			} else {
			for (var i=0,j=materials.length; i<j; i++) {
				if ( hideMaterials.includes(materials[i].name) ) {
						colorString = "1,1,1,1";
		color = colorString.split(',')
					.map(numberString => parseFloat(numberString));
          modelViewer.model.materials[i].alphaMode = 'MASK'
		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
			}
			}}});'

[atrox.zip](https://github.com/google/model-viewer/files/7663699/atrox.zip)

**Reply by slightHYP** [reply]
I used the setter function to update the alphaMode value. Something like this:

`modelViewer.model.materials[i].setAlphaMode("MASK");`

In my case this works well. I have tried your .glb model and I'm able to hide it.

**Reply by dmkrtz** [reply]
`modelViewer.model.materials[i].setAlphaMode("MASK");`

This little piece of code made it possible for me to finally use >1.8.0. 

```
function hideTexture(that, i) {
	var colorString = ""; 
	var color = "";

	if (that.innerHTML == '<i class="fas fa-eye"></i>') {	
		colorString = "0,0,0,0";
		colorBefore = modelViewer.model.materials[i].pbrMetallicRoughness.baseColorFactor;
		alphaModeBefore = modelViewer.model.materials[i].getAlphaMode();

		color = colorString.split(',')
			.map(numberString => parseFloat(numberString));
		that.innerHTML = '<i class="fas fa-eye-slash"></i>'
		$(that).attr("colorBefore", colorBefore);
		$(that).attr("alphaModeBefore", alphaModeBefore);
		
		modelViewer.model.materials[i].setAlphaMode("MASK");
		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
	} else {
		colorString = $(that).attr("colorBefore");
		alphaModeBefore = $(that).attr("alphaModeBefore");

		color = colorString.split(',')
			.map(numberString => parseFloat(numberString));
		that.innerHTML = '<i class="fas fa-eye"></i>'

		modelViewer.model.materials[i].setAlphaMode(alphaModeBefore);
		modelViewer.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);
	}
}
```

Now that I add the MASK mode to the material before changing color, the `0, 0, 0, 0` does its job. Not all materials of a model I provide in my program need any Alpha Mode except Opaque, so this is possibly the best way to achieve that.

Thank you very much @slightHYP!!

**Reply by Biglow230** [reply]
> I used the setter function to update the alphaMode value. Something like this:
> 
> `modelViewer.model.materials[i].setAlphaMode("MASK");`
> 
> In my case this works well. I have tried your .glb model and I'm able to hide it.

I'm still rather confused as to why I am having such a difficult time getting this to work.

I even tried @dmkrtz updated method. I just can't seem to get this to do what's being shown and I'm confused on where the disconnect would be.

The updated code is as follows: Sorry in advance if I'm asking dumb questions 😔 I'm still very new to this level of coding. Thank you for your time an patience. 
`Show Hide <input  type="checkbox" id="checked"  >`


```
modelViewer.addEventListener('load',()=>{
        
var isChecked = $(this).prop("checked");
var materials = modelViewer.model.materials;
const hideMaterials = [ "Material0GMObj0.820" ];

if(isChecked) {
for (var i=0,j=materials.length; i<j; i++) {
	if ( hideMaterials.includes(materials[i].name) ) {
	
		colorString = "1,1,1,0";
		color = colorString.split(',')
	.map(numberString => parseFloat(numberString));
         modelViewer.model.materials[i].setAlphaMode("MASK");
	modelViewer.model.materials[i].pbrMetallicRoughness.sourceObject.setBaseColorFactor(color);
			}
			}
			} else {
	for (var i=0,j=materials.length; i<j; i++) {
	if ( hideMaterials.includes(materials[i].name) ) {
		colorString = "1,1,1,1";
		color = colorString.split(',')
	.map(numberString => parseFloat(numberString));
        modelViewer.model.materials[i].setAlphaMode("MASK");
	modelViewer.model.materials[i].pbrMetallicRoughness.sourceObject.setBaseColorFactor(color);
			}
			}}});
```

**Reply by slightHYP** [reply]
I've just tried your 3d model and this loop works for me:

        for (var i=0,j=materials.length; i<j; i++) {
            if ( hideMaterials.includes(materials[i].name) ) {
		        colorString = "1,1,1,0";
		        color = colorString.split(',').map(numberString => parseFloat(numberString));
                    this.modelViewerElement.model.materials[i].setAlphaMode("MASK");
	                **this.modelViewerElement.model.materials[i].pbrMetallicRoughness.setBaseColorFactor(color);**
	        }
        }


The baseColorFactor have to be applied directly to the material not to the sourceObject.

**Reply by Biglow230** [reply]
@slightHYP Thank you so much for your patience, used this exact code and it still isn't working. Apparently I may have overlooked that "colorString" apparently is not defined but wouldn't it be defined by the "1,1,1,0" or am I missing something here? I most likely am.

**Reply by slightHYP** [reply]
The setBaseColorFactor function accepts an array of 4 integer and in your code is correct, as you can see in the following gif.

![model-viewer-hiding-materials](https://user-images.githubusercontent.com/12815732/145540677-ca04a386-98e4-4689-8ff4-23872c32a22a.gif)

So the problem is somewhere else.

Is the isChecked flag correct?
You retrieve the value inside the 'load' callback function with the following code:

`$(this).prop("checked");`

You are in another scope. I suppose that the `$(this)` doesn't refer to your checkbox state, it returns always "false" and skip the "hiding loop" entering to the "else" condition where colorString is "1,1,1,1". Could this be?

Have a look there.
